net: posix: fix hostname config conflict with unique updates

PragatiGarg-eaton opened the pull request:
## Overview

Fixes Kconfig conflict between POSIX networking and unique hostname updates that was introduced in v3.7.1.

## Problem

Enabling `CONFIG_POSIX_API=y` auto-selects `CONFIG_POSIX_NETWORKING=y`, which force-selects `CONFIG_NET_HOSTNAME_DYNAMIC=y`. This conflicts with `CONFIG_NET_HOSTNAME_UNIQUE_UPDATE=y`, preventing the use of POSIX networking with unique hostname updates.

## Solution

Decouple `NET_HOSTNAME_MAX_LEN` from `NET_HOSTNAME_DYNAMIC` dependency:
- Remove forced selection of `NET_HOSTNAME_DYNAMIC` from POSIX networking
- Make hostname length configuration available independently
- Add build assertions - Ensure proper validation at compile time

## Testing

- [x] Verified fix resolves Kconfig conflicts
- [x] Validated unique hostname update operations
- [x] Confirmed hostname length constants are properly defined
- [x] Build tests pass for various configuration combinations

Fixes #95811

cfriedt commented:
@PragatiGarg-eaton - tests need to pass in order for the change to be merged. Can you revisit?

Please feel free to jump on discord to ask for pointers for running failing tests.

PragatiGarg-eaton committed:
net: posix: fix hostname config conflict with unique updates

Decouple NET_HOSTNAME_MAX_LEN from NET_HOSTNAME_DYNAMIC to resolve
Kconfig conflict when using POSIX networking with unique hostname
updates. This restores functionality that worked in Zephyr 3.6.

- Remove forced selection of NET_HOSTNAME_DYNAMIC from POSIX_NETWORKING
- Make NET_HOSTNAME_MAX_LEN available independently
- Add build assertions for hostname length validation
- Update test configurations for new dependency structure

Fixes: zephyrproject-rtos#95811
Signed-off-by: Pragati Garg <pragatigarg@eaton.com>

cfriedt commented:
I'll see if I can isolate the issues that is causing tests to fail and suggest a couple of changes.

PragatiGarg-eaton commented:
> I'll see if I can isolate the issues that is causing tests to fail and suggest a couple of changes

Thanks @cfriedt, I am also checking in the meanwhile. Per my analysis, The problem is with CONFIG_NET_HOSTNAME_UNIQUE_UPDATE=y, the hostname update happens asynchronously, so the MAC in the hostname may not match the interface's current MAC yet. 
When UNIQUE_UPDATE is enabled, do you suggest to update the test case to:
1. Wait 100ms for async update to complete
2. Re-fetch hostname
3. Parse MAC again
4. Then verify it matches

PragatiGarg-eaton commented:
> @cfriedt , Please review the PR, all the twister jobs were passed in the last run.

Hey @cfriedt gentle reminder for the PR review, please have a look whenever you have time.

cfriedt reviewed (dismissed):
LGTM - @jukkar and @rlubos should review as well.

PragatiGarg-eaton commented:
@jukkar @rlubos, could you please check if the changes look good to merge?

PragatiGarg-eaton commented:
@jukkar, @cfriedt  if everything looks good on your end and all approvals are in, can we go ahead with the merge please?

pdgendt commented:
> @jukkar, @cfriedt if everything looks good on your end and all approvals are in, can we go ahead with the merge please?

We're in the feature freeze for 4.4, as this probably isn't a release blocker bugfix, it'll probably need to go in after.

PragatiGarg-eaton commented:
> > @jukkar, @cfriedt if everything looks good on your end and all approvals are in, can we go ahead with the merge please?
> 
> We're in the feature freeze for 4.4, as this probably isn't a release blocker bugfix, it'll probably need to go in after.

okay thanks for the confirmation 🙌

PragatiGarg-eaton commented:
@cfriedt, @pdgendt, @jukkar can you please trigger the checks and approve the PR again. I've rebased and resolved the conflicts post v4.4 release.

PragatiGarg-eaton commented:
> @cfriedt, @pdgendt, @jukkar can you please trigger the checks and approve the PR again. I've rebased and resolved the conflicts post v4.4 release.

Hi @cfriedt A few documentation-related checks are currently failing, and I’m not entirely sure about the root cause. Could you please help me understand what needs to be addressed next?

PragatiGarg-eaton commented:
> > @cfriedt, @pdgendt, @jukkar can you please trigger the checks and approve the PR again. I've rebased and resolved the conflicts post v4.4 release.
> 
> Hi @cfriedt A few documentation-related checks are currently failing, and I’m not entirely sure about the root cause. Could you please help me understand what needs to be addressed next?

Hello @cfriedt @pdgendt @jukkar, can you please help me with this issue, or if these are the false failures, can you please retrigget the checks?
